backend/routes/userNews.js

```diff
@@ -161,12 +173,104 @@ router.put("/:newsId/bookmarked", async (req, res) => {
 
       personalNews.push(newArticle[0]);
     }
+
     res.json(personalNews);
   } catch (error) {
     res.status(500).json({ error: "Error updating metadata" });
   }
 });
 
+// get a personalized feed
+router.post("/personalized", async (req, res) => {
+    const newNews = await prisma.userNewsCache.findMany({
+      where: { userId: req.session.userId },
+      orderBy: {
+        id: "desc",
+      },
+    });
+
+    const personalNews = [];
+    // there are entries found
+    for (const art of newNews) {
+      const newArticle = await prisma.news.findMany({
+        where: { id: art.newsId },
+      });
+
+      newArticle[0].score = art.score;
+      newArticle[0].bookmarked = art.bookmarked;
+
+      personalNews.push(newArticle[0]);
+    }
+
+    try {
+         // sort news based on engagement scores
+    personalNews.sort((a, b) =>  (b['score']) - (a['score']))
+    const top10 = personalNews.slice(0,10); // top 10 articles with highest engagement scores
+
+    let categories = []   // aggregate the categories present in the top 10
+    for (const article of top10) {
+        for (const category of article.category) {
+            if (!categories.includes(category)) {
+                categories.push(category)
+            }
+        }
+    }
+
+    // perform similiarity analysis
+    weightedScores = {} // scores for every category (how many points a news article gets for having this category)
+    for (article of top10) {
```

Comment:
We've discussed this already but noting for posterity, for the next iteration we want to use more signals than just categories